pkg/mutation/mutators/assignimage/imageparser.go

```diff
@@ -0,0 +1,135 @@
+
+	// domainRegexp defines a schema for a domain component.
+	domainRegexp = regexp.MustCompile(`(^\w[\w\-_]*\.[\w\-_\.]*[\w](:\d+)?$)|(^localhost(:\d+)?$)`)
+
+	// pathRegexp defines a schema for a location component. It follows the convention
+	// specified in the docker distribution reference. The regex  restricts
+	// location-components to start with an alphanumeric character, with following
+	// parts able to be separated by a separator (one period, one or two
+	// underscore and multiple dashes).
+	pathRegexp = regexp.MustCompile(`^[a-z0-9]+(?:(?:(?:[._/]|__|[-]*)[a-z0-9]+)+)?`)
+
+	// tagRegexp defines a schema for a tag component. It must start with `:` or `@`.
+	tagRegexp = regexp.MustCompile(`(^:[\w][\w.-]{0,127}$)|(^@[A-Za-z][A-Za-z0-9]*([-_+.][A-Za-z][A-Za-z0-9]*)*[:][0-9A-Fa-f]{32,}$)`)
+)
+
+type image struct {
+	domain string
+	path   string
+	tag    string
+}
+
+func mutateImage(domain, path, tag, mutableImgRef string) string {
+	oldImg := newImage(mutableImgRef)
+	newImg := oldImg.newMutatedImage(domain, path, tag)
+	return newImg.fullRef()
+}
+
+func newImage(imageRef string) image {
+	domain, remainder := splitDomain(imageRef)
+	var pt string
+	tag := ""
+	if tagSep := strings.IndexAny(remainder, ":@"); tagSep > -1 {
+		pt = remainder[:tagSep]
+		tag = remainder[tagSep:]
+	} else {
+		pt = remainder
+	}
+
+	return image{domain: domain, path: pt, tag: tag}
+}
+
+func (img image) newMutatedImage(domain, path, tag string) image {
+	return image{
+		domain: ignoreUnset(img.domain, domain),
+		path:   ignoreUnset(img.path, path),
+		tag:    ignoreUnset(img.tag, tag),
+	}
+}
+
+// ignoreUnset returns `new` if `new` is set, otherwise it returns `old`.
+func ignoreUnset(old, new string) string {
+	if new != "" {
+		return new
+	}
+	return old
+}
+
+func (img image) fullRef() string {
+	domain := img.domain
+	if domain != "" {
+		domain += "/"
+	}
+	return domain + img.path + img.tag
+}
+
+func splitDomain(name string) (domain, remainder string) {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
+		return "", name
+	}
+	return name[:i], name[i+1:]
+}
+
+func validateDomain(domain string) error {
+	if domain == "" {
+		return nil
+	}
+
+	if !domainRegexp.MatchString(domain) {
+		return fmt.Errorf("assignDomain %q does not match pattern %s", domain, domainRegexp.String())
+	}
+
+	if d, r := splitDomain(domain + "/"); d != domain || r != "" {
+		return fmt.Errorf("domain %q could not be regognized as a valid domain", domain)
+	}
+
+	return nil
+}
+
+func validateImageParts(domain, path, tag string) error {
+	if domain == "" && path == "" && tag == "" {
+		return fmt.Errorf("at least one of [assignDomain, assignPath, assignTag] must be set")
+	}
+	if err := validateDomain(domain); err != nil {
+		return err
+	}
+	// match the whole string for path (anchoring with `$` is tricky here)
+	if path != "" && path != pathRegexp.FindString(path) {
```

Comment:
Similar to what we did with domain, we should ensure that path doesn't include a tag. We can separate out

```
	tag := ""
	if tagSep := strings.IndexAny(remainder, ":@"); tagSep > -1 {
		pt = remainder[:tagSep]
		tag = remainder[tagSep:]
	} else {
		pt = remainder
	}
```

into its own function to avoid duplicating that code.

I realize the regex precludes this, this is future proofing.

Reply:
Done. Great suggestion!

Reply:
I see us validating tag can round trip, but are we validating that path does not contain a tag?

E.g. if I feed path into `splitTag`, will I get a tag?

Reply:
I added a check in `validateTag()` to make sure `splitTag()` does not return a path component, like so:
```
if p, t := splitTag(tag); t != tag || p != "" { 
    return fmt.Errorf(...)
}
```
Is this what you had in mind?